webrender/src/device/gfx/buffer.rs

```diff
@@ -268,93 +272,120 @@ impl<B: hal::Backend> InstancePoolBuffer<B> {
+            buffers: Vec::new(),
             alignment_mask,
             non_coherent_atom_size_mask,
-            current_buffer_index: 0,
+            next_buffer_index: 0,
         }
     }
 
     pub(super) fn add<T: Copy>(
         &mut self,
         device: &B::Device,
-        mut data: &[T],
+        mut instance_data: &[T],
         heaps: &mut Heaps<B>,
-    ) {
-        assert_eq!(self.data_stride, mem::size_of::<T>());
-        while !data.is_empty() {
-            if self.current_buffer().buffer.buffer_size
-                == self.current_buffer().offset * self.data_stride
-            {
-                self.current_buffer_index += 1;
-                if self.buffers.len() <= self.current_buffer_index {
-                    self.buffers.push(InstancePoolBuffer::new(
+        free_buffers: &mut SmallVec<[InstancePoolBuffer<B>; 16]>,
+    ) -> std::ops::Range<usize> {
+        fn instance_data_to_u8_slice<T: Copy>(data: &[T]) -> &[u8] {
+            unsafe {
+                std::slice::from_raw_parts(
+                    data.as_ptr() as *const u8,
+                    data.len() * mem::size_of::<T>(),
+                )
+            }
+        }
+
+        let data_stride = mem::size_of::<T>();
+        let mut range = 0..0;
+        let mut first_iteration = true;
+        while !instance_data.is_empty() {
+            let need_new_buffer = self.buffers.is_empty()
+                || !self.current_buffer().can_store_data(data_stride);
+            if need_new_buffer {
+                let buffer = match free_buffers.pop() {
+                    Some(b) => b,
+                    None => InstancePoolBuffer::new(
                         device,
                         heaps,
                         hal::buffer::Usage::VERTEX,
-                        self.data_stride,
                         self.alignment_mask,
                         self.non_coherent_atom_size_mask,
-                    ))
-                }
-            }
-
-            let update_size = if (self.current_buffer().offset + data.len()) * self.data_stride
-                > self.current_buffer().buffer.buffer_size
-            {
-                self.current_buffer().buffer.buffer_size / self.data_stride
-                    - self.current_buffer().offset
+                    ),
+                };
+                self.buffers.push(buffer);
+                self.next_buffer_index += 1;
             } else {
-                data.len()
-            };
-
-            self.buffers[self.current_buffer_index].update(device, &data[0 .. update_size]);
-
-            data = &data[update_size ..]
+                self.buffers[self.next_buffer_index - 1].align_offset_to(data_stride);
+            }
+            if first_iteration {
+                range.start = self.next_buffer_index - 1;
+                first_iteration = false;
+            }
+            let update_size = (self.current_buffer().free() / data_stride).min(instance_data.len());
```

Comment:
the usage of `current_buffer()` seems inconsistent: it's mixed up with raw `self.buffers[...]` accesses

Reply:
Those are mutable borrows, so I've added `current_buffer_mut()` for those cases: https://github.com/szeged/webrender/pull/300/commits/d5c90a5aaa484f69dfdbbddacc74449a7ffa6b60#diff-5dbe2f57bb9273f9a363bcb71ec6834eR357